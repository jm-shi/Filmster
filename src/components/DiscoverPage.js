import React from 'react';
import axios from 'axios';
import { history } from '../AppRouter';
import { NavLink } from 'react-router-dom';

import LoadingPage from './LoadingPage';
import Gallery from './Gallery';
import Navbar from './Navbar';

class DiscoverPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      results: [],
      loading: true
    };
    this.getMovies = this.getMovies.bind(this);
  }

  getMovies(props) {
    let searchType = props.match.params.type;
    searchType = searchType === undefined ? 'now_playing' : searchType;
    const api = `https://api.themoviedb.org/3/movie/${searchType}?api_key=${
      process.env.REACT_APP_MOVIE_API_KEY
    }&language=en-US&page=1`;

    axios
      .get(api)
      .then(response => {
        this.setState(() => ({
          results: response.data.results,
          loading: false
        }));
      })
      .catch(error => {
        console.log(error.message);
      });
  }

  componentDidMount() {
    this.getMovies(this.props);
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    this.getMovies(nextProps);
  }

  render() {
    const { loading } = this.state;
    return (
      <div>
        <Navbar />

        {loading ? <LoadingPage /> : null}

        <div className="container container--row container--margin-top-ms">
          <div className="grid">
            <NavLink
              exact
              activeClassName="selected"
              className="link link__nav"
              to="/discover"
            >
              Now Playing
            </NavLink>
            <NavLink
              activeClassName="selected"
              className="link link__nav"
              to="/discover/popular"
            >
              Popular
            </NavLink>
            <NavLink
              activeClassName="selected"
              className="link link__nav"
              to="/discover/top_rated"
            >
              Top Rated
            </NavLink>
            <NavLink
              activeClassName="selected"
              className="link link__nav"
              to="/discover/upcoming"
            >
              Upcoming
            </NavLink>
          </div>
        </div>

        <Gallery history={history} movies={this.state.results} />
      </div>
    );
  }
}

export default DiscoverPage;
