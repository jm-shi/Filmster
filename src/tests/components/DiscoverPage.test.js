import React from 'react';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { shallow } from 'enzyme';
import DiscoverPage from '../../components/DiscoverPage';

let mockAdapter, request;
const props = {
  match: {
    params: {
      type: 'now_playing'
    }
  }
};

describe('DiscoverPage component', () => {
  beforeEach(() => {
    mockAdapter = new MockAdapter(axios);
    request = `https://api.themoviedb.org/3/movie/now_playing?api_key=${
      process.env.REACT_APP_MOVIE_API_KEY
    }&language=en-US&page=1`;
    mockAdapter.onGet(request).reply(200, {
      data: {
        results: [],
        loading: false
      }
    });
  });

  afterEach(() => {
    mockAdapter.reset();
  });

  afterAll(() => {
    mock.restore();
  });

  it('should render DiscoverPage', () => {
    const wrapper = shallow(<DiscoverPage {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle componentDidMount', async () => {
    const spy = jest.spyOn(DiscoverPage.prototype, 'getMovies');
    const wrapper = shallow(<DiscoverPage {...props} />);
    await wrapper.instance().componentDidMount();
    expect(spy).toHaveBeenCalled();
  });

  it('should handle componentWillReceiveProps', async () => {
    const spy = jest.spyOn(DiscoverPage.prototype, 'getMovies');
    const wrapper = shallow(<DiscoverPage {...props} />);
    await wrapper.instance().UNSAFE_componentWillReceiveProps(props);
    expect(spy).toHaveBeenCalled();
  });
});
