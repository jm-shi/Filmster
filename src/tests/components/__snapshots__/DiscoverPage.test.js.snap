// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscoverPage component should render DiscoverPage 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DiscoverPage
    match={
        Object {
            "params": Object {
              "type": "now_playing",
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Navbar />,
        <LoadingPage />,
        <div
          className="container container--row container--margin-top-ms"
>
          <div
                    className="grid"
          >
                    <NavLink
                              activeClassName="selected"
                              aria-current="page"
                              className="link link__nav"
                              exact={true}
                              to="/discover"
                    >
                              Now Playing
                    </NavLink>
                    <NavLink
                              activeClassName="selected"
                              aria-current="page"
                              className="link link__nav"
                              to="/discover/popular"
                    >
                              Popular
                    </NavLink>
                    <NavLink
                              activeClassName="selected"
                              aria-current="page"
                              className="link link__nav"
                              to="/discover/top_rated"
                    >
                              Top Rated
                    </NavLink>
                    <NavLink
                              activeClassName="selected"
                              aria-current="page"
                              className="link link__nav"
                              to="/discover/upcoming"
                    >
                              Upcoming
                    </NavLink>
          </div>
</div>,
        <Gallery
          history={
                    Object {
                              "action": "POP",
                              "block": [Function],
                              "createHref": [Function],
                              "go": [Function],
                              "goBack": [Function],
                              "goForward": [Function],
                              "length": 1,
                              "listen": [Function],
                              "location": Object {
                                "hash": "",
                                "pathname": "/",
                                "search": "",
                                "state": undefined,
                              },
                              "push": [Function],
                              "replace": [Function],
                            }
          }
          movies={Array []}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="grid"
>
            <NavLink
                        activeClassName="selected"
                        aria-current="page"
                        className="link link__nav"
                        exact={true}
                        to="/discover"
            >
                        Now Playing
            </NavLink>
            <NavLink
                        activeClassName="selected"
                        aria-current="page"
                        className="link link__nav"
                        to="/discover/popular"
            >
                        Popular
            </NavLink>
            <NavLink
                        activeClassName="selected"
                        aria-current="page"
                        className="link link__nav"
                        to="/discover/top_rated"
            >
                        Top Rated
            </NavLink>
            <NavLink
                        activeClassName="selected"
                        aria-current="page"
                        className="link link__nav"
                        to="/discover/upcoming"
            >
                        Upcoming
            </NavLink>
</div>,
          "className": "container container--row container--margin-top-ms",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <NavLink
                activeClassName="selected"
                aria-current="page"
                className="link link__nav"
                exact={true}
                to="/discover"
>
                Now Playing
</NavLink>,
              <NavLink
                activeClassName="selected"
                aria-current="page"
                className="link link__nav"
                to="/discover/popular"
>
                Popular
</NavLink>,
              <NavLink
                activeClassName="selected"
                aria-current="page"
                className="link link__nav"
                to="/discover/top_rated"
>
                Top Rated
</NavLink>,
              <NavLink
                activeClassName="selected"
                aria-current="page"
                className="link link__nav"
                to="/discover/upcoming"
>
                Upcoming
</NavLink>,
            ],
            "className": "grid",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "activeClassName": "selected",
                "aria-current": "page",
                "children": "Now Playing",
                "className": "link link__nav",
                "exact": true,
                "to": "/discover",
              },
              "ref": null,
              "rendered": "Now Playing",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "activeClassName": "selected",
                "aria-current": "page",
                "children": "Popular",
                "className": "link link__nav",
                "to": "/discover/popular",
              },
              "ref": null,
              "rendered": "Popular",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "activeClassName": "selected",
                "aria-current": "page",
                "children": "Top Rated",
                "className": "link link__nav",
                "to": "/discover/top_rated",
              },
              "ref": null,
              "rendered": "Top Rated",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "activeClassName": "selected",
                "aria-current": "page",
                "children": "Upcoming",
                "className": "link link__nav",
                "to": "/discover/upcoming",
              },
              "ref": null,
              "rendered": "Upcoming",
              "type": [Function],
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "history": Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          },
          "movies": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Navbar />,
          <LoadingPage />,
          <div
            className="container container--row container--margin-top-ms"
>
            <div
                        className="grid"
            >
                        <NavLink
                                    activeClassName="selected"
                                    aria-current="page"
                                    className="link link__nav"
                                    exact={true}
                                    to="/discover"
                        >
                                    Now Playing
                        </NavLink>
                        <NavLink
                                    activeClassName="selected"
                                    aria-current="page"
                                    className="link link__nav"
                                    to="/discover/popular"
                        >
                                    Popular
                        </NavLink>
                        <NavLink
                                    activeClassName="selected"
                                    aria-current="page"
                                    className="link link__nav"
                                    to="/discover/top_rated"
                        >
                                    Top Rated
                        </NavLink>
                        <NavLink
                                    activeClassName="selected"
                                    aria-current="page"
                                    className="link link__nav"
                                    to="/discover/upcoming"
                        >
                                    Upcoming
                        </NavLink>
            </div>
</div>,
          <Gallery
            history={
                        Object {
                                    "action": "POP",
                                    "block": [Function],
                                    "createHref": [Function],
                                    "go": [Function],
                                    "goBack": [Function],
                                    "goForward": [Function],
                                    "length": 1,
                                    "listen": [Function],
                                    "location": Object {
                                      "hash": "",
                                      "pathname": "/",
                                      "search": "",
                                      "state": undefined,
                                    },
                                    "push": [Function],
                                    "replace": [Function],
                                  }
            }
            movies={Array []}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="grid"
>
              <NavLink
                            activeClassName="selected"
                            aria-current="page"
                            className="link link__nav"
                            exact={true}
                            to="/discover"
              >
                            Now Playing
              </NavLink>
              <NavLink
                            activeClassName="selected"
                            aria-current="page"
                            className="link link__nav"
                            to="/discover/popular"
              >
                            Popular
              </NavLink>
              <NavLink
                            activeClassName="selected"
                            aria-current="page"
                            className="link link__nav"
                            to="/discover/top_rated"
              >
                            Top Rated
              </NavLink>
              <NavLink
                            activeClassName="selected"
                            aria-current="page"
                            className="link link__nav"
                            to="/discover/upcoming"
              >
                            Upcoming
              </NavLink>
</div>,
            "className": "container container--row container--margin-top-ms",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <NavLink
                  activeClassName="selected"
                  aria-current="page"
                  className="link link__nav"
                  exact={true}
                  to="/discover"
>
                  Now Playing
</NavLink>,
                <NavLink
                  activeClassName="selected"
                  aria-current="page"
                  className="link link__nav"
                  to="/discover/popular"
>
                  Popular
</NavLink>,
                <NavLink
                  activeClassName="selected"
                  aria-current="page"
                  className="link link__nav"
                  to="/discover/top_rated"
>
                  Top Rated
</NavLink>,
                <NavLink
                  activeClassName="selected"
                  aria-current="page"
                  className="link link__nav"
                  to="/discover/upcoming"
>
                  Upcoming
</NavLink>,
              ],
              "className": "grid",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "activeClassName": "selected",
                  "aria-current": "page",
                  "children": "Now Playing",
                  "className": "link link__nav",
                  "exact": true,
                  "to": "/discover",
                },
                "ref": null,
                "rendered": "Now Playing",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "activeClassName": "selected",
                  "aria-current": "page",
                  "children": "Popular",
                  "className": "link link__nav",
                  "to": "/discover/popular",
                },
                "ref": null,
                "rendered": "Popular",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "activeClassName": "selected",
                  "aria-current": "page",
                  "children": "Top Rated",
                  "className": "link link__nav",
                  "to": "/discover/top_rated",
                },
                "ref": null,
                "rendered": "Top Rated",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "activeClassName": "selected",
                  "aria-current": "page",
                  "children": "Upcoming",
                  "className": "link link__nav",
                  "to": "/discover/upcoming",
                },
                "ref": null,
                "rendered": "Upcoming",
                "type": [Function],
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "history": Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            },
            "movies": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
